AArch64: Enable VM snapshot/restore on AArch64

MrXinWang opened the pull request:
This pull request enables the VM snapshot/restore functionality on AArch64.

Fixes: https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1644

Signed-off-by: Henry Wang <Henry.Wang@arm.com>

rbradford reviewed (commented):
Thanks! This looks fine but I think @liuw should look for hypervisor changes and @sboeuf for general snapshot/restore

MrXinWang commented on hypervisor/src/kvm/mod.rs:
```diff
@@ -860,9 +1116,22 @@ impl cpu::Vcpu for KvmVcpu {
+        let mut state = CpuState::default();
+        // Get this vCPUs multiprocessing state.
+        state.mp_state = self.get_mp_state()?;
+
```
Hmmm yes. I will fix this together with other commits from @liuw and @sboeuf, hope this would be ok to you :)

MrXinWang replied:
Done

MrXinWang commented on hypervisor/src/kvm/mod.rs:
```diff
@@ -946,10 +1215,17 @@ impl cpu::Vcpu for KvmVcpu {
     #[cfg(target_arch = "aarch64")]
     fn set_state(&self, state: &CpuState) -> cpu::Result<()> {
-        warn!("CPU state was not restored");
+        self.restore_core_registers(&state.core_regs)?;
```
same as above.

MrXinWang replied:
Done

liuw commented on hypervisor/src/cpu.rs:
```diff
@@ -295,6 +322,37 @@ pub trait Vcpu: Send + Sync {
+    /// Restore the state of the core registers.
+    ///
+    #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+    fn restore_core_registers(&self, state: &kvm_bindings::kvm_regs) -> Result<()>;
```
This trait is not KVM specific so I would rather not see any KVM specific code here.

I think you can use `StandardRegisters` here.

MrXinWang replied:
Thanks for pointing this out @liuw. Indeed I did not think of the Hyper-V stuff for the hypervisor crate. I've changed the `kvm_regs` to `StandardRegisters`. Hope this would address your concerns.

liuw commented on hypervisor/src/cpu.rs:
```diff
@@ -295,6 +322,37 @@ pub trait Vcpu: Send + Sync {
+    /// Save the state of the system registers.
+    ///
+    #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+    fn save_system_registers(&self, state: &mut Vec<kvm_one_reg>) -> Result<()>;
```
Same here. Not sure if there is a suitable export name that matches `kvm_one_reg` yet. If not, maybe we can invent one.

MrXinWang replied:
@liuw Same as above. I've changed the `kvm_one_reg` to an invented `OneRegister`.

sboeuf reviewed (commented):
Nice PR @MrXinWang.
I have a few comments in the code, and an additional one is that you should be able to enable the snapshot/restore test for Aarch64 in the CI now. This should prove the whole PR works correctly.
On my side, I'm gonna do a few tests with snapshot/restore to verify nothing is breaking the behavior for x86_64.

MrXinWang commented on hypervisor/src/cpu.rs:
```diff
@@ -300,6 +322,32 @@ pub trait Vcpu: Send + Sync {
+    /// Save the state of the core registers.
+    ///
+    #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+    fn save_core_registers(&self, state: &mut kvm_regs) -> Result<()>;
```
Fixed this one.

MrXinWang commented on vmm/src/seccomp_filters.rs:
```diff
@@ -106,6 +106,13 @@ const KVM_GET_ONE_REG: u64 = 0x4010_aeab;
+#[cfg(target_arch = "aarch64")]
+pub const SYS_FTRUNCATE: libc::c_long = 46;
+#[cfg(target_arch = "x86_64")]
+pub const SYS_FTRUNCATE: libc::c_long = 77;
```
Fixed this one in commit 51729882746e30bdd81bd6a877f6cf9eb6b9101f

MrXinWang commented on vmm/src/device_manager.rs:
```diff
@@ -1161,6 +1163,41 @@ impl DeviceManager {
         self.gic_device_entity.as_ref()
     }
+    #[cfg(target_arch = "aarch64")]
+    pub fn construct_gicr_typers(&self, vcpu_states: &[CpuState]) {
```
Hi @sboeuf. Thanks very much for the comment! Actually I hope I would have a discussion about the design about the GIC before I do any modifications, as I am also not sure current code design makes very good sense:

My original thought would be: In current design for AArch64, the software-implemented [devices/src/gic](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/devices/src/gic.rs) is only to provide interrupt delivery service, and the real vGIC device is created through the ioctl here: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/5a3d54fa1945b8c45122e1af33382c7337d869f0/arch/src/aarch64/mod.rs#L153

When we need to save VM, we need to save/restore the vGIC, and I think based on the current design of the VM structure https://github.com/cloud-hypervisor/cloud-hypervisor/blob/5a3d54fa1945b8c45122e1af33382c7337d869f0/vmm/src/vm.rs#L288 we can only access the real vGIC device through the device_manager. This is the reason why the `gic_device_entity` field and related methods (`get/set_gic_device_entity`, `construct_gicr_typers`) were introduced in commit 8e0713a2810f7339afd50b928c99343ecac294e1.

I did think of implement the `Snapshottable` trait for GIC in [devices/src/gic](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/devices/src/gic.rs). However later I noticed that for GICv2, GICv3 and GICv3ITS, the registers (or states) that need to be saved are different. So I think the `Snapshottable` trait should be implemented separately in arch/src/gic. As the real vGIC device is created using trait object, we need to downcast the trait object to specific type. 

Hopefully now things is more clear, any suggestions about how we can optimise the design?

sboeuf replied:
Oh I understand now. Well then if only the device manager can access it, and also because this is a common piece of code for all GIC devices, let's keep it where it is :)
Thanks for the explanation!

MrXinWang commented on vmm/src/vm.rs:
```diff
@@ -1476,6 +1505,59 @@ impl Snapshottable for Vm {
         }
 
+        #[cfg(target_arch = "aarch64")]
+        {
```
Yes that definitely makes sense, however as this part is only for saving the vGIC state. I am not very sure if naming it to something like `add_vGIC_snapshot_section` and put it in vm.rs is proper...Is this ok to you?

sboeuf replied:
That's fine :)

MrXinWang commented on vmm/src/vm.rs:
```diff
@@ -1425,6 +1430,30 @@ impl Snapshottable for Vm {
         vm_snapshot.add_snapshot(self.memory_manager.lock().unwrap().snapshot()?);
+
+        #[cfg(target_arch = "aarch64")]
+        {
```
same as above. Probably we can name it to `restore_vGIC_snapshot_session`?

sboeuf replied:
Same here, fine with me.

MrXinWang commented:
Hi @rbradford @liuw @sboeuf - Thanks for reviewing! I believe I have addressed all of the comments :) (sorry about the `git push -f`, cannot think a cleaner way to update the code...)

> an additional one is that you should be able to enable the snapshot/restore test for Aarch64 in the CI now. This should prove the whole PR works correctly.

Thanks @sboeuf, the CI commit 3349d3a65a23ad7c7d1ceee6109bb9670c0cd87f has been added. I tested on both x86_64 and AArch64, nothing in CI is broken.

Also one thing I need to mention is that in the updated commit 3880736bf7dd08d2ecd6921ecc787c6bffc0fde9, I made a small change when creating vCPUs in VM restore: this change guarantees that the restored vCPUs are created from vCPU0 to vCPUn in order. See [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1704/commits/3880736bf7dd08d2ecd6921ecc787c6bffc0fde9#diff-39805f31fede77dcaa77b9055d65e821R1506).

The reason of this change is that: currently the `snapshots` field is implemented by a hashmap:
https://github.com/cloud-hypervisor/cloud-hypervisor/blob/036c2e5e45f362409b4bb92c17073ae5d03efe60/vm-migration/src/lib.rs#L78
we don't have a specific order when we call `iter()` in restoring vCPU. On AArch64, since the vGIC states are bonded to vCPUs (GICv3 has redistributor and CPU interface) through the [MPIDR](https://developer.arm.com/docs/ddi0595/h/aarch64-system-registers/mpidr_el1) register, we need to make sure that the restored VM has the exactly same vCPUs in the same order as those in the booting process. Although x86_64 would not need to be such strict...this change does not break x86_64.

sboeuf commented:
@MrXinWang 
>The reason of this change is that: currently the snapshots field is implemented by a hashmap

Yes unfortunately the iterator from a `HashMap` is not ordered correctly. Do you think switching to a `BTreeMap` would still work? I'm asking because if there is no blocker in using a `BTreeMap` instead of a `HashMap`, we might get an ordered iterator for free.

MrXinWang commented:
@sboeuf I also thought this! If changing `HashMap` to `BTreeMap` is ok with you, I will try and update the code :)

MrXinWang commented:
Hi @sboeuf, `BTreeMap` works perfectly :) Thanks!

sboeuf commented:
>If changing HashMap to BTreeMap is ok with you, I will try and update the code :)

I'm fine with it. Unless @sameo or @rbradford have any objection, you can go ahead :)

MrXinWang commented:
@sboeuf Commit 6139d487ba481828a9e8509c2e5d550cf3b7a9ce updated with `BTreeMap` code. Thanks :)

sameo commented:
> > If changing HashMap to BTreeMap is ok with you, I will try and update the code :)
> 
> I'm fine with it. Unless @sameo or @rbradford have any objection, you can go ahead :)

It's fine with me. `BTreeMap` is basically an ordered by key map, which is what we need here. It should not break anything.

sboeuf reviewed (approved):
Looks good to me.
@rbradford @liuw PTAL

michael2012z reviewed (approved):
LGTM. 
Thanks.

rbradford reviewed (approved):
rslgtm - needs a rebase though

MrXinWang commented:
Hi @rbradford. Rebased. Thanks :)
